File: teamapps-client/teamapps-client-core-components/src/main/ts/component/tool-container/toolbar/ToolbarButtonGroup.ts
/*-
 * ========================LICENSE_START=================================
 * TeamApps
 * ---
 * Copyright (C) 2014 - 2022 TeamApps.org
 * ---
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =========================LICENSE_END==================================
 */
import {DtoToolbarButtonGroup as DtoToolbarButtonGroup} from "../../../generated/DtoToolbarButtonGroup";
import {OrderedDictionary} from "../../../util/OrderedDictionary";
import {ToolbarButton} from "./ToolbarButton";
import {Component, insertAfter, insertBefore, parseHtml, TeamAppsEvent, TeamAppsUiContext} from "teamapps-client-core";

import {DtoToolbarButton as DtoToolbarButton} from "../../../generated/DtoToolbarButton";
import {ButtonVisibilities, Toolbar} from "./Toolbar";
import {DtoDropDownButtonClickInfo} from "../../../generated";
import {outerWidthIncludingMargins} from "../../../Common";

export class ToolbarButtonGroup {
	public readonly onButtonClicked: TeamAppsEvent<{buttonId: string, dropDownButtonClickInfo: DtoDropDownButtonClickInfo}> = new TeamAppsEvent();

	private config: DtoToolbarButtonGroup;
	private visible: boolean = true;
	private $buttonGroupWrapper: HTMLElement;
	private $buttonGroup: HTMLElement;
	private buttons: OrderedDictionary<ToolbarButton>;
	private $separator: HTMLElement;
	private buttonsShiftedToOverflowDropDown: ToolbarButton[] = [];

	constructor(buttonGroupConfig: DtoToolbarButtonGroup, private toolbar: Toolbar) {
		const $buttonGroupWrapper = parseHtml('<div class="button-group-wrapper"></div>');

		const $buttonGroup = parseHtml(`<div class="toolbar-button-group"></div>`);
		$buttonGroupWrapper.appendChild($buttonGroup);

		this.config = buttonGroupConfig;
		this.$buttonGroupWrapper = $buttonGroupWrapper;
		this.$buttonGroup = $buttonGroup;
		this.buttons = new OrderedDictionary<ToolbarButton>();

		for (let j = 0; j < buttonGroupConfig.buttons.length; j++) {
			const buttonConfig: DtoToolbarButton = buttonGroupConfig.buttons[j];
			const button = this.createButton(buttonConfig);
			this.buttons.push(buttonConfig.buttonId, button);
			$buttonGroup.append(button.getMainDomElement());
		}

		if (buttonGroupConfig.showGroupSeparator) {
			this.$separator = parseHtml('<div class="toolbar-group-separator">');
			$buttonGroupWrapper.append(this.$separator);
		}

		this.setVisible(buttonGroupConfig.visible);


	}

	public getId() {
		return this.config.groupId;
	}

	public get position() {
		return this.config.position;
	}

	private createButton(buttonConfig: DtoToolbarButton) {
		const button = new ToolbarButton(buttonConfig);
		button.onClicked.addListener(dropDownButtonClickInfo => this.onButtonClicked.fire({buttonId: buttonConfig.buttonId, dropDownButtonClickInfo}));
		return button;
	}

	public setDropDownComponent(buttonId: string, component: Component) {
		this.buttons.getValue(buttonId).setDropDownComponent(component);
	}

	public setButtonVisible(buttonId: string, visible: boolean) {
		const button = this.buttons.getValue(buttonId);
		if (button) {
			button.setVisible(visible);
			this.updateVisibility();
		}
	}

	public addButton(buttonConfig: DtoToolbarButton, neighborButtonId: string, beforeNeighbor: boolean) {
		const button = this.createButton(buttonConfig);

		const existingButton = this.buttons.getValue(buttonConfig.buttonId);
		if (existingButton) {
			this.removeButton(buttonConfig.buttonId);
		}

		const neighborButton = this.buttons.getValue(neighborButtonId);
		if (neighborButton) {
			if (beforeNeighbor) {
				insertBefore(button.getMainDomElement(), neighborButton.getMainDomElement());
				this.buttons.insertBeforeValue(buttonConfig.buttonId, button, neighborButton);
			} else {
				insertAfter(button.getMainDomElement(), neighborButton.getMainDomElement());
				this.buttons.insertAfterValue(buttonConfig.buttonId, button, neighborButton);
			}
		} else {
			this.$buttonGroup.appendChild(button.getMainDomElement());
			this.buttons.push(buttonConfig.buttonId, button);
		}

		this.updateVisibility();
	}

	public removeButton(buttonId: string): void {
		const button = this.buttons.getValue(buttonId);

		if (button) {
			this.buttons.remove(buttonId);
			button.getMainDomElement().remove();
		}

		// TODO destroy dropdown

		this.updateVisibility();
	}

	public setVisible(visible: boolean) {
		this.visible = visible;
		this.updateVisibility();
	}

	public isVisible() {
		return this.visible;
	}

	private updateVisibility() {
		let hasVisibleButton = this.buttons.values.some(button => {
			return button.isVisible() && this.buttonsShiftedToOverflowDropDown.indexOf(button) === -1;
		});
		this.$buttonGroupWrapper.classList.toggle("pseudo-hidden", !(this.visible && hasVisibleButton));
	}

	setButtonColors(buttonId: string, backgroundColor: string, hoverBackgroundColor: string) {
		this.buttons.getValue(buttonId).setColors(backgroundColor, hoverBackgroundColor);
	}

	public calculateButtonVisibilities(availableWidth: number): ButtonVisibilities {
		let info: ButtonVisibilities = {
			fittingButtons: [],
			nonFittingButtons: [],
			hiddenButtons: []
		};
		let usedWidth = this.$separator ? outerWidthIncludingMargins(this.$separator) : 0;
		this.buttons.values.forEach((button) => {
			if (this.visible && button.isVisible()) {
				if (usedWidth + button.optimizedWidth <= availableWidth) {
					this.buttonsShiftedToOverflowDropDown = this.buttonsShiftedToOverflowDropDown.filter(b => b !== button);
					info.fittingButtons.push({groupId: this.config.groupId, buttonId: button.id, button});
				} else {
					this.buttonsShiftedToOverflowDropDown.push(button);
					info.nonFittingButtons.push({groupId: this.config.groupId, buttonId: button.id, button});
				}
				usedWidth += button.optimizedWidth;
			} else {
				this.buttonsShiftedToOverflowDropDown = this.buttonsShiftedToOverflowDropDown.filter(b => b !== button);
				info.hiddenButtons.push({groupId: this.config.groupId, buttonId: button.id, button});
			}
		});
		this.updateVisibility();
		return info;
	}

	public getMainDomElement(): HTMLElement {
		return this.$buttonGroupWrapper;
	}

	public getButtonConfigs() {
		return this.buttons.values.map(b => b.config);
	}

	setButtonHasDropDown(buttonId: string, hasDropDown: boolean) {
		this.buttons.getValue(buttonId).setHasDropDown(hasDropDown);
	}

	public getConfig() {
		return this.config;
	}

	closeDropDown(buttonId: string) {
		this.buttons.getValue(buttonId)?.closeDropDown();
	}
}

File: client/projector-client-core-components/src/main/java/org/teamapps/projector/component/core/toolbar/ToolbarButtonGroup.java
/*-
 * ========================LICENSE_START=================================
 * TeamApps
 * ---
 * Copyright (C) 2014 - 2022 TeamApps.org
 * ---
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =========================LICENSE_END==================================
 */
package org.teamapps.projector.component.core.toolbar;

import org.teamapps.projector.common.format.Color;
import org.teamapps.projector.component.Component;
import org.teamapps.projector.dataextraction.PropertyExtractor;
import org.teamapps.projector.dataextraction.PropertyProvider;
import org.teamapps.projector.template.Template;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

public class ToolbarButtonGroup implements Comparable<ToolbarButtonGroup> {

	private AbstractToolContainer toolContainer;
	private String clientId = UUID.randomUUID().toString();

	private final List<ToolbarButton> buttons = new ArrayList<>();
	private ToolbarButtonGroupPosition position;
	private boolean rightSide;

	private boolean visible = true;
	private boolean showGroupSeparator = true;

	private Template buttonTemplate; // if null, will fallback to toolbar's valueExtractor
	private PropertyProvider propertyProvider; // if null, will fallback to toolbar's valueExtractor

	public ToolbarButtonGroup() {
		this(new ArrayList<>(), ToolbarButtonGroupPosition.CENTER);
	}

	public ToolbarButtonGroup(ToolbarButtonGroupPosition position) {
		this(new ArrayList<>(), position);
	}

	public ToolbarButtonGroup(List<ToolbarButton> buttons, ToolbarButtonGroupPosition position) {
		buttons.forEach(this::addButton);
		this.position = position;
	}

	public void setToolContainer(AbstractToolContainer toolContainer) {
		this.toolContainer = toolContainer;
	}

	public void setButtons(List<ToolbarButton> buttons) {
		new ArrayList<>(this.buttons).forEach(button -> removeButton(button));
		buttons.forEach(button -> addButton(button));
	}

	public ToolbarButton addButton(ToolbarButton button) {
		return addButton(button, null, false);
	}

	public ToolbarButton addButton(ToolbarButton button, ToolbarButton neighborButton, boolean beforeNeighbor) {
		buttons.add(button);
		button.setToolbarButtonGroup(this);
		if (toolContainer != null) {
			toolContainer.handleAddButton(this, button, neighborButton != null ? neighborButton.getClientId() : null, beforeNeighbor);
		}
		return button;
	}

	public void removeButton(ToolbarButton button) {
		buttons.remove(button);
		button.setToolbarButtonGroup(null);
		if (toolContainer != null) {
			toolContainer.handleButtonRemoved(this, button);
		}
	}

	protected String getClientId() {
		return clientId;
	}

	public List<ToolbarButton> getButtons() {
		return buttons;
	}

	public ToolbarButtonGroupPosition getPosition() {
		return position;
	}

	public DtoToolbarButtonGroup createDtoToolbarButtonGroup() {
		List<DtoToolbarButton> buttons = this.buttons.stream()
				.map(button -> button.createDtoToolbarButton())
				.collect(Collectors.toList());
		DtoToolbarButtonGroup buttonGroup = new DtoToolbarButtonGroup(clientId, buttons);
		buttonGroup.setVisible(visible);
		buttonGroup.setShowGroupSeparator(showGroupSeparator);
		return buttonGroup;
	}

	public AbstractToolContainer getToolContainer() {
		return toolContainer;
	}

	public boolean isVisible() {
		return visible;
	}

	public void setVisible(boolean visible) {
		boolean oldValue = this.visible;
		this.visible = visible;
		if (oldValue != visible && this.toolContainer != null) {
			toolContainer.handleGroupVisibilityChange(this.clientId, visible);
		}
	}

	public boolean isShowGroupSeparator() {
		return showGroupSeparator;
	}

	public void setShowGroupSeparator(boolean showGroupSeparator) {
		this.showGroupSeparator = showGroupSeparator;
	}

	public void setPosition(ToolbarButtonGroupPosition position) {
		this.position = position;
	}

	@Override
	public int compareTo(ToolbarButtonGroup o) {
		return position.compareTo(o.getPosition());
	}

	/*package-private*/ void handleButtonVisibilityChange(String buttonClientId, boolean visible) {
		if (this.toolContainer != null) {
			this.toolContainer.handleButtonVisibilityChange(this.clientId, buttonClientId, visible);
		}
	}

	/*package-private*/ void handleDropDownComponentUpdate(ToolbarButton button, Component component) {
		if (this.toolContainer != null) {
			this.toolContainer.handleButtonSetDropDownComponent(this, button, component);
		}
	}

	/*package-private*/ void handleCloseDropdown(ToolbarButton button) {
		if (this.toolContainer != null) {
			this.toolContainer.handleCloseDropdown(this, button);
		}
	}

	public void handleColorChange(String buttonClientId, Color backgroundColor, Color hoverBackgroundColor) {
		if (this.toolContainer != null) {
			this.toolContainer.handleButtonColorChange(this.clientId, buttonClientId, backgroundColor, hoverBackgroundColor);
		}
	}

	public Template getButtonTemplate() {
		return buttonTemplate;
	}

	public void setButtonTemplate(Template buttonTemplate) {
		this.buttonTemplate = buttonTemplate;
	}

	public Template getAppliedTemplate() {
		return this.buttonTemplate != null ? this.buttonTemplate : this.toolContainer.getButtonTemplate();
	}

	public PropertyProvider getPropertyProvider() {
		return propertyProvider;
	}

	public void setPropertyProvider(PropertyProvider propertyProvider) {
		this.propertyProvider = propertyProvider;
	}

	public void setPropertyExtractor(PropertyExtractor propertyExtractor) {
		this.setPropertyProvider(propertyExtractor);
	}
	public PropertyProvider getAppliedPropertyProvider() {
		return propertyProvider != null ? propertyProvider : toolContainer.getPropertyProvider();
	}

	public void setClientId(String clientId) {
		this.clientId = clientId;
	}

	public boolean isRightSide() {
		return rightSide;
	}

	public void setRightSide(boolean rightSide) {
		this.rightSide = rightSide;
	}
}
